test/Microsoft.AspNetCore.Authentication.Test/OpenIdConnect/OpenIdConnectEventTests.cs

```diff
@@ -28,1579 +28,938 @@ namespace Microsoft.AspNetCore.Authentication.Test.OpenIdConnect
-                };
-                events.OnUserInformationReceived = context =>
-                {
-                    userInfoReceived = true;
-                    throw new NotImplementedException("TestException");
-                };
-                events.OnAuthenticationFailed = context =>
-                {
-                    authFailed = true;
-                    Assert.Equal("TestException", context.Exception.Message);
-                    context.Fail("Authentication was aborted from user code.");
-                    return Task.FromResult(0);
-                };
-                events.OnRemoteFailure = context =>
-                {
-                    remoteFailure = true;
-                    return Task.FromResult(0);
-                };
-            }),
-            context =>
+            var events = new ExpectedOidcEvents()
+            {
+                ExpectMessageReceived = true,
+                ExpectTokenValidated = true,
+                ExpectAuthorizationCodeReceived = true,
+                ExpectTokenResponseReceived = true,
+                ExpectUserInfoReceived = true,
+                ExpectAuthenticationFailed = true,
+                ExpectRemoteFailure = true,
+            };
+            events.OnUserInformationReceived = context =>
             {
-                return context.Response.WriteAsync(context.Request.Path);
-            });
+                events.InvokedUserInfoReceived = true;
+                throw new NotImplementedException("TestException");
+            };
+            events.OnAuthenticationFailed = context =>
+            {
+                events.InvokeAuthenticationFailed = true;
+                Assert.Equal("TestException", context.Exception.Message);
+                context.Fail("Authentication was aborted from user code.");
+                return Task.FromResult(0);
+            };
+            var server = CreateServer(events, AppWritePath);
 
             var exception = await Assert.ThrowsAsync<Exception>(delegate
             {
                 return PostAsync(server, "signin-oidc", "id_token=my_id_token&state=protected_state&code=my_code");
             });
 
             Assert.Equal("Authentication was aborted from user code.", exception.InnerException.Message);
-
-            Assert.True(messageReceived);
-            Assert.True(tokenValidated);
-            Assert.True(codeReceived);
-            Assert.True(tokenResponseReceived);
-            Assert.True(userInfoReceived);
-            Assert.True(authFailed);
-            Assert.True(remoteFailure);
+            events.ValidateExpectations();
         }
 
         [Fact]
         public async Task OnAuthenticationFailed_HandledWithoutTicket_NoMoreEventsRun()
         {
-            var messageReceived = false;
-            var tokenValidated = false;
-            var codeReceived = false;
-            var tokenResponseReceived = false;
-            var userInfoReceived = false;
-            var a
```

Comment:
Nit: I can't tell if there's a common set of these which you are setting to true in all the tests, but maybe its worth adding a helper to slim these down to something more like:

var events = ExpectedOidcEvents.CreateDefaultExpectations();
// only set the one or two test specific ones. (which hopefully is none for many tests?)

Reply:
It's progressive. You start with only one and work your way though until you've hit almost all of them.  There's no subset for all tests.